akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -250,6 +250,12 @@ import scala.collection.JavaConverters._
         true
       } else false
 
+    private def tryCancel(): Boolean =
+      if (activeSubstreamsMap.isEmpty || (activeSubstreamsMap.size == substreamWaitingToBePushed.size)) {
```

Comment:
That new condition is quite subtle, why is it needed? It might be clearer if we clear `substreamWaitingToBePushed` in `onDownstreamFinish` and remove it from the `activeSubstreamsMap` as well. Otherwise, please add a comment explaining the condition.

Reply:
It is needed because a new key can be pending (because downstream backpressured) before the downstream cancels.

How could we remove it from `activeSubstreamsMap`? Maybe keep the key on a `var`? I can do that, or I can add a comment. WDYT?